gnome-initial-setup/pages/language/cc-language-chooser.c

```diff
@@ -145,10 +147,91 @@ language_widget_new (const char *locale_id,
         return widget->box;
 }
 
+typedef struct {
+        CcLanguageChooser *chooser;
+        GtkWidget         *row;
+} ScrollData;
+
+static inline gint
+get_selected_language_row_y (GtkWidget *row)
+{
+        GtkAllocation alloc;
+
+        gtk_widget_get_allocation (row, &alloc);
+
+        return alloc.y + alloc.height / 2.0;
+}
+
+static gboolean
+update_scroll_position (gpointer user_data)
+{
+        CcLanguageChooser *self;
+        GtkAdjustment *vadjustment;
+        ScrollData *data;
+        GtkWidget *row;
+        gdouble value, real_value;
+        gdouble max, page, page_increment;
+
+        data = user_data;
+        self = data->chooser;
+        row = data->row;
+
+        /* While the scrolled window is not mapped, we keep waiting */
+        if (!gtk_widget_get_mapped (self->priv->language_scrolledwindow))
+                return G_SOURCE_CONTINUE;
+
+        vadjustment = gtk_scrolled_window_get_vadjustment (GTK_SCROLLED_WINDOW (self->priv->language_scrolledwindow));
+        page = gtk_adjustment_get_page_size (vadjustment);
+        max = gtk_adjustment_get_upper (vadjustment);
+
+        page_increment = gtk_adjustment_get_page_increment (vadjustment);
+        real_value = get_selected_language_row_y (row) - page_increment / 2.0;
+        value = gtk_adjustment_get_value (vadjustment);
+
+        gtk_adjustment_set_page_increment (vadjustment, real_value - value);
+
+        g_signal_emit_by_name (self->priv->language_scrolledwindow,
+                               "scroll-child",
+                               GTK_SCROLL_PAGE_FORWARD,
+                               FALSE);
+
+        gtk_adjustment_set_page_increment (vadjustment, page_increment);
+
+out:
+        self->priv->scroll_timeout_id = 0;
+        return G_SOURCE_REMOVE;
+}
+
+static void
+schedule_scroll (CcLanguageChooser *self,
+                 GtkWidget         *row)
+{
+        ScrollData *data;
+
+        data = g_new0 (ScrollData, 1);
+        data->row = row;
+        data->chooser = self;
+
+        /* Nothing is scheduled, we should do it now */
+        if (self->priv->scroll_timeout_id > 0)
+                g_source_remove (self->priv->scroll_timeout_id);
```

Comment:
Perhaps because the condition you want to check is whether the widget is realized instead? I am not very comfortable merging the timeout hack without at least understanding why we would be forced to use it.